Nominatim Geocoder Locator filter - add countryCodes configuration option

uprel opened the pull request:
This PR adds Configuration option to Nominatim Geocoder Locator filter for filtering results to one or more countries. (Two letter country codes, comma separated). Configuration value is optional, if left blank it will work as now.

[Nominatim Geocoder](https://github.com/qgis/QGIS/blob/f38d190f56f691ace1ddb779a4597c6aefb0ec7f/src/core/geocoding/qgsnominatimgeocoder.cpp#L36) already supports countryCodes parameter so only GUI part and linking setting to request was missing.
[Test ](https://github.com/qgis/QGIS/blob/f38d190f56f691ace1ddb779a4597c6aefb0ec7f/tests/src/python/test_qgsnominatimgeocoder.py#L70) also supports this parameter so no new tests are created for this.
This change does not effect Batch Nominatim Geocoder processing tool.

Please backport to 3.42 and 3.40
Needs documentation

![image](https://github.com/user-attachments/assets/8bb3cffb-d8ae-48b2-9ed7-1be8b1210641)
![image](https://github.com/user-attachments/assets/27c08941-0df5-46f6-b9a9-72b166e990a3)

arx-it committed:
Distinguish Arcgis feature/map icons in the browser

benoitdm-oslandia committed:
fix(qgisserver): suppress wait in fcgi response destructor by using timed_mutex

benoitdm-oslandia committed:
clang-tidy: remove clang-analyzer-optin.cplusplus.VirtualCall warning in QgsFcgiServerResponse

benoitdm-oslandia committed:
chore: no need to fail if directory exists

benoitdm-oslandia committed:
fix(qgisserver): add select call to avoid to wait on recv when socket is closed

benoitdm-oslandia committed:
fix(fcgiserver): log warning message instead of debug

benoitdm-oslandia committed:
fix(qgisserver): add more feedback checkpoints

Djedouas committed:
fix(layoutlegendwidget): correctly move legend nodes when they are filtered

Fixes #60383.

In a layout, when moving a legend node on a filtered legend, the
computed offset was wrong if some nodes were hidden between the node to
move and its destination.

The offset is now computed with the next/previous visible index.

uprel committed:
Add country codes config to Nominatim Geocoder Locator filter

uprel committed:
Remove debug line

uprel commented:
Backport also to 3.42

alexbruy commented on src/app/locator/qgsnominatimlocatorfilter.cpp:
```diff
@@ -50,3 +52,30 @@ void QgsNominatimLocatorFilter::triggerResult( const QgsLocatorResult &result )
+
+  // Load existing settings
+  QgsSettings settings;
+  countryCodesEdit->setText( settings.value( "locator_filters/nominatim_geocoder/country_codes", "", QgsSettings::App ).toString() );
```
Maybe use new settings tree API here and in other places?

uprel replied:
Don't know, I used the code from other Locator filters as example. Like this https://github.com/qgis/QGIS/blob/f38d190f56f691ace1ddb779a4597c6aefb0ec7f/src/app/locator/qgsalllayersfeatureslocatorfilter.cpp#L219
I can check this, but not sure what will be better. Is approach I used deprecated?

alexbruy replied:
It is not deprecated, but IMHO for new code we should use new API. Also with this API new setting will be nicely exposed in the advanced settings editor.

uprel replied:
I see what you mean now, yes it makes sense for new code to use new setting. I will look at that, but first trying to fix override position of fetchResults below...

uprel replied:
I now see also documentation to use [settingsTree](https://docs.qgis.org/3.40/en/docs/developers_guide/codingstandards.html#settings), so thanks for pointing out and I should have check this before.

uprel committed:
Update src/app/locator/qgsnominatimlocatorfilter.cpp

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

uprel committed:
Update src/app/locator/qgsnominatimlocatorfilter.cpp

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

uprel committed:
revert qgsabstractgeocoderlocatorfilter.cpp from master

alexbruy committed:
port Define Projection algorithm to C++

uprel committed:
- Fix formLayout to gridLayout
- override fetchResults but doesn't get called. Help needed!

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

uprel commented on src/core/geocoding/qgsabstractgeocoderlocatorfilter.cpp:
```diff
@@ -45,10 +47,26 @@ QString QgsAbstractGeocoderLocatorFilter::prefix() const
+    QString countryCodes = settings.value( "locator_filters/nominatim_geocoder/country_codes", "", QgsSettings::App ).toString().trimmed();
+
+    nominatimGeocoder->setCountryCodes( countryCodes );
+  }
```
I know, but override doesn't work. I commited this override for reference, it doesn't get called. I always get fetchResults from QgsAbstractGeocoderLocatorFilter. Overriding hierarchy in this case is somehow weird. Do you see easy fix that I missed?

uprel replied:
OK, now I moved fetchResults to QgsGeocoderLocatorFilter because I was able to override it there. If possible I would leave it like this, unless someone shows me what's missing to override it in QgsNominatimLocatorFilter. I will also change settings to use settingsTree

nyalldawson replied:
> I know, but override doesn't work.

Show your code please, I can't diagnose "doesn't work" :rofl: !

uprel replied:
Of course, sorry. So now commited code contains fetchResults override in QgsGeocoderLocatorFilter and that works. If I put it in QgsNominatimLocatorFilter

**src/app/locator/qgsnominatimlocatorfilter.cpp**
```
void QgsNominatimLocatorFilter::fetchResults(const QString &string, const QgsLocatorContext &context, QgsFeedback *feedback)
{
  QgsSettings settings;
  QString countryCodes = settings.value( "locator_filters/nominatim_geocoder/country_codes", "", QgsSettings::App ).toString().trimmed();

  QgsNominatimGeocoder *nominatimGeocoder = dynamic_cast<QgsNominatimGeocoder *>( geocoder() );
  nominatimGeocoder->setCountryCodes( countryCodes );

  QgsAbstractGeocoderLocatorFilter::fetchResults( string, context, feedback);
}
```

**src/app/locator/qgsnominatimlocatorfilter.h**
` void fetchResults( const QString &string, const QgsLocatorContext &context, QgsFeedback *feedback ) override;`

This code does not get executed, regarding if I add override method in its parent QgsGeocoderLocatorFilter or not.

When debugging I see a problem that here in qgslocator.cpp:
https://github.com/qgis/QGIS/blob/70ad90ef8480da2158c55d8b77b6e2ff2ea0ca39/src/core/locator/qgslocator.cpp#L239
filter is type QgsGeocoderLocatorFilter and not expected QgsNominatimLocatorFilter.

So then it's logical that doesn't get to fetchResults in QgsNominatimLocatorFilter. But I don't know why is that, since filter is registered as QgsNominatimLocatorFilter, here in qgisapp.cpp:
https://github.com/qgis/QGIS/blob/70ad90ef8480da2158c55d8b77b6e2ff2ea0ca39/src/app/qgisapp.cpp#L4106

nyalldawson replied:
See https://github.com/qgis/QGIS/pull/60793

uprel replied:
Great, thanks for solving this

uprel committed:
Move fetchResults from NominatimLocatorFilter to GeocodeLocatorFilter

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nyalldawson committed:
Add missing clone method to QgsNominatimLocatorFilter

Fixes missing accreditation message when triggering results

uprel committed:
revert qgsgeocoderlocatorfilter.cpp from master

uprel committed:
override fetchResults to add countryCodes from settings

uprel committed:
Move new setting country-codes to QgsSettingsTree

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

uprel committed:
remove blank line for test_sipfiles to work

troopa81 committed:
fix(Oracle): Synchronize test connection button state with OK button

wonder-sk committed:
Large scene support: shift scene's origin if the camera got too far

This should greatly improve our capability to handle large scenes (e.g. more than 50km across)

Also add a checkbox to the debug panel to allow devs temporarily turn it off.

Djedouas committed:
fix(clang-tidy): use unnamed namespace instead of static keyword

For "private" functions like the underscore-prefixed-by-convention
functions, use an unnamed namespace to affect visibility over static
keyword which has many meanings.

uprel commented:
Kindly asking @alexbruy and @nyalldawson to comment on changes or merge this. Failed test is not related to this code. Thanks!

wonder-sk committed:
Review from Stefanos + test fix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

uclaros committed:
Avoid crashing when a texture is provided by an extension

uclaros committed:
Don't assume all flat polygons face up

wonder-sk committed:
Fix reset view functionality

Relax the assumption that origin is always in the center of the scene

wonder-sk committed:
Fix tests - call setExtent(), not setOrigin()

wonder-sk committed:
setting origin is not required (it is done in setExtent() anyway)

wonder-sk committed:
Fix 3d map scene test

Origin may get updated in Qgs3DMapScene constructor, so let's set
ours once the Qgs3DMapScene is constructed

Alex-Kent committed:
When duplicating a layer copy layer tree layer's custom properties [fixes #60884]

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

uclaros committed:
Treat flat polygons as always facing up when extruding

nyalldawson committed:
Add more flexible API variant for QgsVectorLayerExport.exportLayer

nyalldawson committed:
Add extent filter for QgsVectorLayerExporter

nyalldawson committed:
Add some missing tests

nyalldawson committed:
Add expression based filter for QgsVectorLayerExporter

nyalldawson committed:
Make map canvas accessible via QgsDataItemGuiContext

Allows browser item GUI operations to access the map canvas

nyalldawson committed:
[feature] Add optional extent, expression based filter when importing
tables to database

Sponsored by City of Canning

nyalldawson committed:
[mssql] Add workaround for incorrect geometry system type returned by sp_describe_first_result_set

Some (versions?) (setups?) of SQL server incorrectly report that
the system type for a geometry columns is "image" in their
sp_describe_first_result_set results.

Let's just automatically override this and set it back to "geometry"
if we know that the column is actually the geometry column.

Fixes broken query layers on these databases

nyalldawson committed:
Fix build, docs

alexbruy committed:
remove unneeded else branch

uprel committed:
revert wrong merge, no change here

uclaros committed:
Adjust profile tests

benoitdm-oslandia committed:
fix(qgisserver): apply review changes

vcloarec committed:
fix confusion between native face and triangle face in 3D mesh layer

nyalldawson committed:
Fix typo

nyalldawson committed:
Nicer appearance for NULL values in execute sql results table

Match appearance to attribute table, showing application NULL
string in a shaded, italic font

nyalldawson committed:
Don't try to show raw binary data in query result table

Commit:
auto sipify 🍺

nyalldawson committed:
Minor decoupling of QgsAggregateMappingModel from QgsFieldMappingModel

nyalldawson committed:
Allow setting native field types for QgsFieldMappingWidget

Previously the widget would always just use a generic hardcoded
list of field types. This isn't appropriate for all destination
providers, which may not support certain field types.

Add API so that an explicit list of supported destination
field types can be set for the widget, restricting the choices
available to users for destination fields.

nyalldawson committed:
Improve field type mapping

Implement fallback logic, so eg if the destination field types
don't support one of the input types exactly, we can still make
reasonable matches to a compatible field

nyalldawson committed:
Add support for setting output fields for QgsVectorLayerExporter

Allows remapping fields from the source fields, including specifying
the exact output field type, and source expression

nyalldawson committed:
[feature] Allow field customisation when importing db tables

This expands the database table import dialog to add a new
field mapping section, which gives users control over the
fields in the created table. Users can rename, set the exact
destination field types, and tweak the source expression
for each output field. Fields can also be excluded from the
import, or new fields created which don't exist in the source
table.

Sponsored by City of Canning

nyalldawson committed:
Remove unused variable

nyalldawson committed:
Update src/3d/qgs3dmapscene.cpp

nyalldawson committed:
Don't allow qgsgeometry_cast to cast away const

Commit:
auto sipify 🍺

uprel commented:
I tried to stash commits into single and this is the result. I will need some help here, what to do to merge this. Failed test is not related.

nyalldawson committed:
Documentation improvements

nyalldawson committed:
Add missing sip downcasts for triangulated surface, polyhedral surface

nyalldawson committed:
Fix qgsgeometry_cast QgsPolyhedralSurface, QgsTriangulatedSurface to base classes

nyalldawson committed:
Expand tests

nyalldawson committed:
Fix typos

Commit:
auto sipify 🍺

Commit:
auto sipify 🍺

uprel commented:
closing, will do another fresh start
